lib/rules/no-constant-condition.js

```diff
@@ -33,7 +35,10 @@ module.exports = {
 
     create(context) {
         const options = context.options[0] || {},
-            checkLoops = options.checkLoops !== false;
+            checkLoops = options.checkLoops !== false,
+            sourceCode = context.getSourceCode();
+
+        let containsYield = false;
```

Comment:
It seems like having a single `containsYield` variable will not be able to handle the case where one loop in the file contains `yield`, and another doesn't.

Additionally, we should make sure the rule handles the case where a `yield` inside a loop is in a function (which would prevent it from pausing the loop). See my other comment with test cases for examples of this.

Instead, my recommendation would be to do the following while traversing the AST:

* Keep of a set of infinite loops for which there hasn't been any `yield` statement found yet.
    * When an infinite loop is entered, add it to the set.
    * When a `yield` is encountered, remove everything from the set.
    * When a loop is exited, report it if it's in the set (and optionally remove it from the set, since it's no longer important).
    * When entering a function, push the current set onto a stack and work with a new empty set.
    * When exiting a function, discard the current set, then pop the old set from the stack and continue working with that one.

By using this strategy, you can always keep track of the current loops that the traverser is inside.

A vaguely similar strategy is used [here](https://github.com/eslint/eslint/blob/7d9e3beeb58c1ee71d53dfcfd3e3b0721dd79b46/lib/rules/object-shorthand.js#L272-L313) and [here](https://github.com/eslint/eslint/blob/7d9e3beeb58c1ee71d53dfcfd3e3b0721dd79b46/lib/rules/object-shorthand.js#L327-L345) in the `object-shorthand` rule. (The goal of that logic is slightly different in that it's trying detect the scope of a given `this` expression within a given arrow function, rather than a `yield` expression within a given loop.)